File: ulina_server/src/discord/commands/create_nation.rs
use crate::database::db;
use crate::discord::commands::shared::{
    continent_option, get_continent, Category, CommandData, CreateCommand, Interaction, OptionType,
};
use crate::discord::helper::Helper;
use crate::discord::ids::{CONTINENT, NAME, USER};
use crate::error::Error;
use crate::get_options;
use serenity::client::Context;

use sqlx::query;

pub const DATA: CommandData = CommandData {
    admin_only: true,
    name: "create-nation",
    category: Category::EditNation,
};

pub fn create(command: &mut CreateCommand) -> &mut CreateCommand {
    command
        .name(DATA.name)
        .description("admin only - create a nation")
        .create_option(|option| {
            option
                .name(NAME)
                .description("name of the new nation")
                .kind(OptionType::String)
                .required(true)
        })
        .create_option(|option| {
            continent_option(option)
                .description("continent of the new nation")
                .required(true)
        })
        .create_option(|option| {
            option
                .name(USER)
                .description("user of the new nation")
                .kind(OptionType::User)
                .required(true)
        })
}

pub async fn create_nation(ctx: &Context, interaction: &Interaction) -> Result<(), Error> {
    let (name, continent, user) = get_options!(
        interaction.data.options,
        NAME,
        String,
        CONTINENT,
        Integer,
        USER,
        User
    );
    let (name, continent, user) = (name?, get_continent(*continent?)?, user?);
    let user = user.id.0.to_string();

    query!(
        "INSERT INTO Nation (continentName, name, removed, ownerDiscord) VALUES (?, ?, false, ?)",
        continent,
        name,
        user
    )
    .execute(db())
    .await?;

    interaction
        .message(&ctx.http, |message| {
            message.content(format!("Successfully created {}", name))
        })
        .await;

    Ok(())
}

File: bot/commands/create-nation.ts
import {SlashCommandBuilder} from '@discordjs/builders';
import {CommandInteraction} from "discord.js";
import {isAdmin, UlinaCommand} from "../bot";
import {CONTINENT_ID, FLAG_ID, NAME_ID, USER_ID} from "../identifiers";
import {continentChoice} from "../nation-edit";
import {allContinents, createNation} from "../../server/database";

module.exports = {
    data: new SlashCommandBuilder()
        .setName('create-nation')
        .setDescription('admin only - create a nation')
        .addStringOption(option =>
            option.setName(NAME_ID)
                .setDescription("name of the new nation").setRequired(true)
        ).addIntegerOption(option =>
            continentChoice(option, "continent of the new nation").setRequired(true)
        ).addUserOption(option =>
            option.setName(USER_ID).setDescription("user of the new nation").setRequired(true)
        ),
    async execute(interaction: CommandInteraction) {
        if (!isAdmin(interaction.member)){
            interaction.reply({content: `you are not an admin.`, ephemeral: true});
            return;
        }
        const name = interaction.options.getString(NAME_ID);
        const continent = allContinents[interaction.options.getInteger(CONTINENT_ID)];
        const ownerDiscord = interaction.options.getUser(USER_ID).id;

        await createNation({name, continent, ownerDiscord, removed: false});
        interaction.reply(`successfully created ${name}`);
    },
    category: "Edit Nation"
} as UlinaCommand;